File: Семинар наставника 2/DZ/QuickSortSeq.cpp
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <ctime>

#define N 100000 // 10000000 1000000 100000 10000
#define ITERS 100 // 10 100

void printArr(int* arr, int n)
{
  for (int i = 0; i < n; i++)
    printf("%d ", arr[i]);
  printf("\n");
}

bool isArraySorted(int* arr, int n)
{
  if (n > 0)
  {
    int prev = arr[0];
    for (int i = 1; i < n; i++)
      if (arr[i] < prev)
        return false;
  }
  return true;
}

void swap(int* a, int* b)
{
  if (*a == *b) return;
  *a = *a + *b;
  *b = *a - *b;
  *a = *a - *b;
}

int partitionLomuto(int* arr, int low, int high)
{
  int pivot = arr[high];
  int i = (low - 1);

  for (int j = low; j < high; j++)
    if (arr[j] <= pivot)
    {
      i++;
      swap(&arr[i], &arr[j]);
    }

  swap(&arr[i + 1], &arr[high]);
  return (i + 1);
}

int partitionHoar(int* arr, int low, int high)
{
  size_t i = low;
  size_t j = high;
  int pivot = arr[(i + j) / 2];
  while (1)
  {
    while (arr[i] < pivot)
      ++i;
    while (arr[j] > pivot)
      --j;

    if (i >= j)
      break;

    swap(&arr[i++], &arr[j--]);
  }
  return j;
}

void quickSortSeq(int* arr, int low, int high)
{
  if (low < high)
  {
    int pi = partitionHoar(arr, low, high);
    // Рекурсивно сортируем элементы, находящиеся до пивота
    quickSortSeq(arr, low, pi);
    // Рекурсивно сортируем элементы, находящиеся после пивота
    quickSortSeq(arr, pi + 1, high);
  }
}

int main()
{
  srand(time(NULL));

  for (int gl = 0; gl < 3; gl ++)
  {
    int* arr = (int*)malloc(N * sizeof(int));
    double time_seq = 0;
    for (int it = 0; it < ITERS; it++)
    {
      // Инициализация массивов случайными значениями
      for (int i = 0; i < N; i++)
      {
        arr[i] = rand() % 100;
        //arr[i] = N - i;
      }
      //printArr(arr, N);

      // Засекаем время перед сортировкой
      double start_time = omp_get_wtime();

      // Сортировка Хоар последовательно
      quickSortSeq(arr, 0, N - 1);

      // Засекаем время после сортировки
      double end_time = omp_get_wtime();
      double elapsed_time = end_time - start_time;
      time_seq += elapsed_time;

      //printArr(arr, N);
      if (!isArraySorted(arr, N))
        printf("Massive is not sorted\n");
    }
    free(arr);
    time_seq /= ITERS;
    printf("Time taken (sequential): %f seconds\n", time_seq);
  }
  
  return 0;
}

File: Семинар наставника 2/DZ/QuickSortSeq.py
import numpy as np
from numba import njit
import time

def isArraySorted(arr):
    n = len(arr)
    if (n > 0):
        prev = arr[0]
        for i in range(n):
            if (arr[i] < prev):
                return False
    return True;

def createRandomArr(n):
    return np.random.randint(0, 100, n)

def quickSortSeq(arr, low, high):
    if (low < high):
        # partitionHoar
        i = low
        j = high
        pivot = arr[(i + j) // 2]
        while (1):
            while (arr[i] < pivot):
                i = i + 1
            while (arr[j] > pivot):
                j = j - 1

            if (i >= j):
                break

            arr[i], arr[j] = arr[j], arr[i]
            i = i + 1
            j = j - 1
        pi = j

        # recurcive call
        quickSortSeq(arr, low, pi)
        quickSortSeq(arr, pi + 1, high)

@njit
def quickSortSeqNjit(arr, low, high):
    if (low < high):
        # partitionHoar
        i = low
        j = high
        pivot = arr[(i + j) // 2]
        while (1):
            while (arr[i] < pivot):
                i = i + 1
            while (arr[j] > pivot):
                j = j - 1

            if (i >= j):
                break

            arr[i], arr[j] = arr[j], arr[i]
            i = i + 1
            j = j - 1
        pi = j

        # recurcive call
        quickSortSeqNjit(arr, low, pi)
        quickSortSeqNjit(arr, pi + 1, high)

@njit
def quickSortIterative(arr):
    stack = []
    stack.append((0, len(arr) - 1))

    while stack:
        low, high = stack.pop()

        if low < high:
            # pivot = partition(arr, low, high)
            # partitionHoar
            i = low
            j = high
            pivot = arr[(i + j) // 2]
            while (1):
                while (arr[i] < pivot):
                    i = i + 1
                while (arr[j] > pivot):
                    j = j - 1

                if (i >= j):
                    break

                arr[i], arr[j] = arr[j], arr[i]
                i = i + 1
                j = j - 1
            pivot = j

            # using stack
            stack.append((low, pivot))
            stack.append((pivot + 1, high))

N = 100000
ITERS = 100

def calcSeqTime():
    arr = createRandomArr(10)
    quickSortSeq(arr, 0, len(arr) - 1)
    full_time = 0
    for it in range(ITERS):
        arr = createRandomArr(N)
        # print(arr)

        start_time = time.time()

        quickSortSeq(arr, 0, len(arr) - 1)

        end_time = time.time()
        elapsed_time = end_time - start_time;
        # print(elapsed_time)
        full_time += elapsed_time

        # print(arr)
        if (not isArraySorted(arr)):
            print("Massive is not sorted\n")

    full_time /= ITERS
    print("Time taken (sequential) without @njit:", full_time, "seconds")

def calcSeqTimeNjit():
    arr = createRandomArr(10)
    quickSortSeqNjit(arr, 0, len(arr) - 1)
    full_time = 0
    for it in range(ITERS):
        arr = createRandomArr(N)
        # print(arr)

        start_time = time.time()

        quickSortSeqNjit(arr, 0, len(arr) - 1)

        end_time = time.time()
        elapsed_time = end_time - start_time;
        # print(elapsed_time)
        full_time += elapsed_time

        # print(arr)
        if (not isArraySorted(arr)):
            print("Massive is not sorted\n")

    full_time /= ITERS
    print("Time taken (sequential) with @njit:", full_time, "seconds")

def calcSeqTimeNjitIterative():
    arr = createRandomArr(10)
    quickSortIterative(arr)
    full_time = 0
    for it in range(ITERS):
        arr = createRandomArr(N)
        # print(arr)

        start_time = time.time()

        quickSortIterative(arr)

        end_time = time.time()
        elapsed_time = end_time - start_time;
        # print(elapsed_time)
        full_time += elapsed_time

        # print(arr)
        if (not isArraySorted(arr)):
            print("Massive is not sorted\n")

    full_time /= ITERS
    print("Time taken (sequential) ITERATIVE with @njit:", full_time, "seconds")

#for i in range(3):
#    calcSeqTime()

for i in range(3):
    calcSeqTimeNjit()

for i in range(3):
    calcSeqTimeNjitIterative()